### PR TITLE
Added SingleStoreDB Vector Store

### PR DESCRIPTION
**Pull request by volodymyr-memsql**
- Added `SingleStoreDB` vector store, which is a wrapper over the SingleStore DB database, that can be used as a vector storage and has an efficient similarity search. 
- Added integration tests for the vector store
- Added jupyter notebook with the example

@dev2049

### CODE REVIEW

File: langchain/vectorstores/singlestoredb.py [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+            vectorstore = SingleStoreDB(
+                embedding_function=embeddings.embed_query,
+            )
+    """
+
+    def _get_connection(self: SingleStoreDB) -> Any:
+        try:
+            import singlestoredb as s2
+        except ImportError:
+            raise ValueError(
```
Comment (leo-gan) [inline langchain/vectorstores/singlestoredb.py:126]: should be
`raise ImportError`

File: tests/integration_tests/vectorstores/test_singlestoredb.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    def embed_query(self, text: str) -> List[float]:
+        return self.normalize(super().embed_query(text))
+
+
+@pytest.fixture
+def texts() -> List[str]:
+    return ["foo", "bar", "baz"]
+
+
+def test_singlestoredb(texts: List[str]) -> None:
```
Comment (leo-gan) [inline tests/integration_tests/vectorstores/test_singlestoredb.py:44]: There can be a check for existing import like this one (I'm just copying code):
```
try:
    from google.cloud import bigquery  # noqa: F401

    bigquery_installed = True
except ImportError:
    bigquery_installed = False


@pytest.mark.skipif(not bigquery_installed, reason="bigquery not installed")
def test_bigquery_loader_no_options() -> None:

```

File: pyproject.toml [hunk 3]
```diff
@@ -102,8 +102,10 @@ azure-cognitiveservices-speech = {version = "^1.28.0", optional = true}
 py-trello = {version = "^0.19.0", optional = true}
 momento = {version = "^1.5.0", optional = true}
 bibtexparser = {version = "^1.4.0", optional = true}
+singlestoredb = {version = "^0.6.1", optional = true}
```
Comment (hwchase17) [inline pyproject.toml:105]: will need to be added to one of the dependencies groups (`all`)

**Comment by volodymyr-memsql** [comment]
@leo-gan, @hwchase17, could you please check it again? Thank you!

File: langchain/vectorstores/base.py [hunk 4]
```diff
@@ -347,6 +358,11 @@ class VectorStoreRetriever(BaseRetriever, BaseModel):
     vectorstore: VectorStore
     search_type: str = "similarity"
     search_kwargs: dict = Field(default_factory=dict)
+    allowed_search_types: ClassVar[Collection[str]] = (
```
Comment (hwchase17) [inline langchain/vectorstores/base.py:361]: id prefer not to do this in this pr - if we do this, should be a separate pr

Comment (hwchase17) [reply langchain/vectorstores/base.py:361]: actually i think probably fine
